### PR TITLE
Add support for osd_delete_sleep configuration value

### PR DESCRIPTION
**Pull request by dzafman**
Fixes: http://tracker.ceph.com/issues/36474

- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/common/options.cc [hunk 1]
```diff
@@ -3380,6 +3380,22 @@ std::vector<Option> get_global_options() {
     .set_default(30)
     .set_description(""),
 
+    Option("osd_delete_sleep", Option::TYPE_FLOAT, Option::LEVEL_ADVANCED)
+    .set_default(0)
+    .set_description("Time in seconds to sleep before next removal transaction (overrides values below)"),
+
+    Option("osd_delete_sleep_hdd", Option::TYPE_FLOAT, Option::LEVEL_ADVANCED)
+    .set_default(5)
```
Comment (dzafman) [inline src/common/options.cc]: 5 seconds for HDD?

File: src/common/options.cc [hunk 2]
```diff
@@ -3380,6 +3380,22 @@ std::vector<Option> get_global_options() {
...
+    .set_description("Time in seconds to sleep before next removal transaction for HDDs"),
+
+    Option("osd_delete_sleep_ssd", Option::TYPE_FLOAT, Option::LEVEL_ADVANCED)
+    .set_default(0)
+    .set_description("Time in seconds to sleep before next removal transaction for SSDs"),
+
+    Option("osd_delete_sleep_hybrid", Option::TYPE_FLOAT, Option::LEVEL_ADVANCED)
+    .set_default(2)
```
Comment (dzafman) [inline src/common/options.cc]: 2 seconds for hybrid?

File: src/common/options.cc [hunk 3]
```diff
@@ -3380,6 +3380,22 @@ std::vector<Option> get_global_options() {
...
+    Option("osd_delete_sleep", Option::TYPE_FLOAT, Option::LEVEL_ADVANCED)
+    .set_default(0)
+    .set_description("Time in seconds to sleep before next removal transaction (overrides values below)"),
+
+    Option("osd_delete_sleep_hdd", Option::TYPE_FLOAT, Option::LEVEL_ADVANCED)
+    .set_default(5)
+    .set_description("Time in seconds to sleep before next removal transaction for HDDs"),
+
+    Option("osd_delete_sleep_ssd", Option::TYPE_FLOAT, Option::LEVEL_ADVANCED)
+    .set_default(0)
```
Comment (dzafman) [inline src/common/options.cc]: No delay for SSD?

File: src/osd/PG.cc [hunk 4]
```diff
@@ -6724,6 +6724,36 @@ void PG::_delete_some(ObjectStore::Transaction *t)
 {
   dout(10) << __func__ << dendl;
 
+  {
+    float osd_delete_sleep = osd->osd->get_osd_delete_sleep();
+    if (osd_delete_sleep > 0 && delete_needs_sleep) {
+      epoch_t e = get_osdmap()->get_epoch();
+      //PGRef pgref(this);
+      auto delete_requeue_callback = new FunctionContext([this, e](int r) {
```
Comment (dzafman) [inline src/osd/PG.cc]: Do I need a _PGRef_ here instead of _this_ for the callback?

File: src/osd/OSD.cc [hunk 5]
```diff
@@ -2569,6 +2569,18 @@ float OSD::get_osd_recovery_sleep()
     return cct->_conf->osd_recovery_sleep_hdd;
 }
 
+float OSD::get_osd_delete_sleep()
+{
+  float osd_delete_sleep = cct->_conf.get_val<double>("osd_delete_sleep");
```
Comment (jdurgin) [inline src/osd/OSD.cc]: if you use the legacy way (cct->_conf->osd_delete_sleep) for these the options will be injectable at runtime

File: src/osd/PG.cc (hunk 4 above)
Comment (jdurgin) [reply src/osd/PG.cc]: yes, iirc that's why it was used in the other sleep code - if the pg gets deleted and then the callback fires, it will crash

Comment (jdurgin) [reply src/osd/PG.cc]: even though this is the pg deletion code, it may still be needed here to handle osd shutdown

File: src/osd/OSD.cc (hunk 5 above)
Comment (dzafman) [reply src/osd/OSD.cc]: @jdurgin I thought that too, but tried it anyway without the legacy options and it worked.

bin/ceph daemon osd.0 config set osd_delete_sleep 60

2018-10-25 16:50:16.521 ... _delete_some Delete scheduled at 2018-10-25 16:51:16.524557

Comment (jdurgin) [reply src/osd/OSD.cc]: oh maybe that's new since the centralized config in mimic

File: src/common/options.cc (hunk 1 above)
Comment (jdurgin) [reply src/common/options.cc]: these defaults seem ok to me. for e.g. 4mb objects, that's a limit of 24mb/s of deletes with the default osd_target_transaction_size of 30 objects, which seems reasonable for HDD.

**Comment by jdurgin** [comment]
for testing this, perhaps you could set the option in one of the thrash configs like we did for snap trim sleep:
```
$ git grep trim.sleep
rados/thrash-old-clients/thrashers/careful.yaml:        osd snap trim sleep: 2
rados/thrash-old-clients/thrashers/default.yaml:        osd snap trim sleep: 2
rados/thrash-old-clients/thrashers/pggrow.yaml:        osd snap trim sleep: .5
rados/thrash/thrashers/careful.yaml:        osd snap trim sleep: 2
rados/thrash/thrashers/default.yaml:        osd snap trim sleep: 2
rados/thrash/thrashers/pggrow.yaml:        osd snap trim sleep: .5
```

**Comment by dzafman** [comment]
Testing passed, all failures seem unrelated.

dzafman-2018-11-06_13:08:30-rados-wip-zafman-testing-24749-distro-basic-smithi
3231266 mon/auth_caps.sh failure
3231188 scrub_test failed due to assert len(objs['inconsistents']) == 1
3231115 http://tracker.ceph.com/issues/36677
3231232 http://tracker.ceph.com/issues/36677

**Pull request merged**